### PR TITLE
Added threadlocaldata implementation for InitGain

### PR DESCRIPTION
**Pull request by uditagar**

### CODE REVIEW

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 1]
```diff
@@ -77,18 +77,19 @@ void InitGain(GGraph& g) {
...
+  for (uint32_t i = 0; i < num_threads; i++) {
+    thread_local_gain_vector.getRemote(i)->resize(g.size() - g.hedges, 0);
   }
-	
+
   uint32_t num_hedges = g.hedges;
 
-	  galois::do_all(
+  galois::do_all(
       galois::iterate(uint32_t{0}, static_cast<uint32_t>(g.hedges)),
```
Comment (nicelhc13) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:93]: it would be very minor, but you can use num_hedges. It would be in the register.

Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:93]: ok, let me do that

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 2]
```diff
@@ -100,16 +114,28 @@ void InitGain(GGraph& g) {
...
+      galois::iterate(num_hedges, static_cast<uint32_t>(g.size())),
+      [&](GNode n) {
+        int gain = 0;
+        for (uint32_t i = 0; i < num_threads; i++) {
+          gain += (*(thread_local_gain_vector.getRemote(i)))[n - num_hedges];
+        }
+
+        g.getData(n).SetPositiveGain(gain);
+      },
+      galois::steal(), galois::loopname("reduce0-1"));
```
Comment (nicelhc13) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:138]: reduce0-1 --> Reduce0-1

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 3]
```diff
@@ -68,17 +68,29 @@ void SortNodesByGainAndWeight(GGraph& graph, std::vector<GNode>& nodes,
...
         node.SetNegativeGain(0);
       },
       galois::loopname("Init-Gains0-0"));
 
+  typedef std::vector<GainTy> LocalGainVector;
+  galois::substrate::PerThreadStorage<LocalGainVector> thread_local_gain_vector;
+
+  uint32_t num_threads = galois::getActiveThreads();
+
+  for (uint32_t i = 0; i < num_threads; i++) {
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:88]: Why is this not parallel?

Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:88]: I have done this now. I will push the latest version in a few minutes

Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:88]: the numbers are with the parallel version

Comment (gurbinder533) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:88]: Cool Cool. Thanks.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 4]
```diff
@@ -100,16 +114,29 @@ void InitGain(GGraph& g) {
...
             }
             // It means that one of p1 or p2 is zero.
             if (nodep == (num_p0_nodes + num_p1_nodes)) {
-              galois::atomicAdd(node_data.GetNegativeGain(), 1);
+              gain_vector[node - num_hedges] -= 1;
             }
           }
         }
       },
       galois::steal(), galois::loopname("Init-Gains0-1"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:126]: what is 0-1 in the name?

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 5]
```diff
@@ -100,16 +114,29 @@ void InitGain(GGraph& g) {
...
+      [&](GNode n) {
+        GainTy gain{0};
+        uint32_t index_n = n - num_hedges;
+        for (uint32_t i = 0; i < num_threads; i++) {
+          gain += (*(thread_local_gain_vector.getRemote(i)))[index_n];
+        }
+
+        g.getData(n).SetPositiveGain(gain);
+      },
+      galois::steal(), galois::loopname("Reduce0-1"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:139]: Same.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp [hunk 6]
```diff
@@ -156,17 +203,41 @@ void InitGain(std::vector<std::pair<uint32_t, uint32_t>>& combined_edgelist,
...
+        }
+
+        auto node_index_pair = combined_nodelist[n];
+        GNode node_id        = node_index_pair.first;
+        uint32_t index       = node_index_pair.second;
+        MetisNode& node_data = g[index]->getData(node_id);
+        node_data.SetPositiveGain(positive_gain);
+        node_data.SetNegativeGain(negative_gain);
+      },
+      galois::steal(), galois::loopname("Reduce1-1"));
```
Comment (gurbinder533) [inline lonestar/eda/cpu/bipart-nested/Helper.cpp:240]: Same.

File: lonestar/eda/cpu/bipart-nested/Helper.cpp (hunk 4 above)
Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:126]: @nicelhc13

Comment (uditagar) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:126]: hochan will do in his PR

Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:126]: Nothing. Just distinguish two InitGains(). Two of them do the same computations. You can suggest better name.

Comment (nicelhc13) [reply lonestar/eda/cpu/bipart-nested/Helper.cpp:126]: OK. It would be better to remove that numbering. we can integrate the timer names.

**Pull request merged**